uavcan/pnp/65534.NodeIDAllocationData.0.1.uavcan

```diff
@@ -0,0 +1,99 @@
+#
+# This message is used for dynamic node ID allocation on all transports except CAN 2.0.
+# For CAN 2.0 there is a dedicated message definition that takes into account the limitations of that transport,
+# making the allocation protocol much more complex.
+#
+# When a node needs to request a node ID dynamically, it will transmit an anonymous message transfer of this type.
+#
+# In order to reduce the probability of CAN ID collisions when multiple nodes are publishing this request over the CAN
+# bus, or any other kind of transport where similar limitations are present, the CAN ID field of the anonymous message
+# transfer includes a "discriminator", which is a special field that has to be filled with pseudo-random data by the
+# transmitting node. Since discriminator collisions are likely to happen, nodes that are requesting dynamic
+# allocations need to be able to handle them correctly. Hence, a collision resolution protocol is defined (like
+# CSMA/CD), which utilizes a randomized request period value referred to as Trequest.
+#
+# The recommended values of the request period should not exceed the maximum defined as a constant below (the minimum is
+# unrestricted). The random interval must be chosen anew per transmission, whereas the discriminator value is allowed
+# to stay constant per node.
+#
+# In the below description the following terms are used:
+# - Allocator - the node that serves allocation requests.
+# - Allocatee - the node that requests an allocation from the Allocator.
+#
+# The response timeout is not explicitly defined for this protocol, as the Allocatee will request a new allocation
+# Trequest units of time later again, unless an allocation has been granted.
+#
+# On the allocatee's side the protocol is defined through the following sets of simple rules:
+#
+#     Rule A. On initialization:
+#     1. The allocatee subscribes to this message.
+#     2. The allocatee starts the Request Timer with a random interval of Trequest.
+#
+#     Rule B. On expiration of the Request Timer:
+#     1. Request Timer restarts with a random interval of Trequest (chosen anew).
+#     2. The allocatee broadcasts an Allocation request message, where the fields are populated as follows:
+#        node_id   - the preferred node ID, or zero if the allocatee doesn't have any specific preference
```

Comment:
I'd like to see two "modes" for this process:

1. the currently described mode where the allocatee accepts any arbitrary node id provided.
2. A mode where the allocator requires that it recognize the provided unique_id. 

The second mode supports vehicle systems that must explicitly map all UIDs to node identifiers. While this may seem to defeat the purpose of the dynamic node allocation it is actually very useful where a given vehicle system is manufactured in large quantities. Imagine an eight wheeled rover where each wheel motor was the same hardware type but the controller had to map wheel motors 0-7 to node identifiers. The manufacturing line can't pre-assign node identifiers since any given motor may end up on any given wheel. Instead it makes sense for the vehicle assembly step to record the UID of each motor and provide the controller with a node-id to UID map. When the system is booted it provides each motor with a pre-determined node id these nodes come on-bus. If any motor is not able to obtain a node id then the map is wrong. It's better to fail to give the motor a node identifier then to give it the wrong identifier (i.e. the controller thinks it's controlling a wheel motor but a gimbal motor got the id instead. Hilarity ensues, cameras are destroyed).

To support this we could document an allocator mode where it always ignored "preferred id" and returned a failure code for "unrecognized UID".

Reply:
Do such impostor nodes even need to know that they are being rejected? The allocator could just ignore them. The allocator will also be aware of the problem. The allocatee, if made aware, wouldn't be able to sensibly react to that error, would it?

Reply:
This is true but it might help a human trying to diagnose why a given wheel motor wasn't passing the "obtain a node id" state. It seems to provide additional diagnostic information without breaking any abstractions.

Reply:
Can't the allocator just emit a log message to the bus? I really don't want to complicate this logic further.

Reply:
Let me ask a different question: how does the allocator know that a given allocation was accepted? For example: 
1. allocatee requests nodeid 1
2. allocator gets the request and queues an allocation message with the nodeid and the allocatee's UID
3. allocatee drops off the bus
4. allocator sends the queued message.

In this state the allocatee did not receive the node id allocation message but the allocator can't know this.

Again, I'm trying to preserve an important use of the current dynamic node allocation service where an allocator is recording the result of each operation to provide a central controller with the system to nodeid map. This new protocol doesn't seem robust enough to trust with this responsibility.

Reply:
> how does the allocator know that a given allocation was accepted?

When the allocator does the actual allocation, it adds a new record to its allocation table (it works the same way for centralized and distributed allocators; in the latter case the act of "adding a new record" involves a bit of shared state synchronization, but the general idea holds and the allocatees are not aware of the difference). If the allocatee falls off the bus (sounds like a roadside accident), the allocator can't be aware of that, but eventually when the allocatee is back online it will request an allocation again. The allocator will see that there is an entry already in the allocation table, and serve the allocatee with the same node ID as before. The whole allocation protocol is entirely idempotent, which makes it robust against the kind of problem you're describing.

> Again, I'm trying to preserve an important use of the current dynamic node allocation service where an allocator is recording the result of each operation to provide a central controller with the system to nodeid map. This new protocol doesn't seem robust enough to trust with this responsibility.

The logic of the protocol is still the same. I just removed an unnecessary follow-up stage to make it simpler and more robust.